pkg/storage/mvcc.go

```diff
@@ -3906,6 +3935,215 @@ func MVCCGarbageCollect(
 	return nil
 }
 
+// MVCCGarbageCollectRanges is similar in functionality to MVCCGarbageCollect but
+// operates on range keys. It does sanity checks that no values exist below
+// range tombstones so that no values are exposed in case point values GC was
+// not performed correctly by level above.
+func MVCCGarbageCollectRanges(
+	ctx context.Context,
+	rw ReadWriter,
+	ms *enginepb.MVCCStats,
+	rangeKeys []roachpb.GCRequest_GCRangeKey,
+	timestamp hlc.Timestamp,
+) error {
+
+	var count int64
+	defer func(begin time.Time) {
+		log.Eventf(ctx, "done with GC evaluation for %d keys at %.2f keys/sec. Deleted %d entries",
+			len(rangeKeys), float64(len(rangeKeys))*1e9/float64(timeutil.Since(begin)), count)
+	}(timeutil.Now())
+
+	// If there are no keys then there is no work.
+	if len(rangeKeys) == 0 {
+		return nil
+	}
+
+	sort.Slice(rangeKeys, func(i, j int) bool {
+		iKey := MVCCRangeKey{StartKey: rangeKeys[i].StartKey, Timestamp: rangeKeys[i].Timestamp}
+		jKey := MVCCRangeKey{StartKey: rangeKeys[j].StartKey, Timestamp: rangeKeys[j].Timestamp}
+		return iKey.Compare(jKey) < 0
+	})
+	// Validate that keys are non-overlapping
```

Comment:
For this to be correct, we'll also need to ensure that `EndKey` > `StartKey`. Consider constructing a set of `MVCCRangeKey` and calling `Validate()` on them to ensure that they're well-formed.